### ISSUE TITLE
Fetch items on demand?

### ISSUE THREAD

**Issue by hidroh** [epishie/Rehash#1]
Hey, nice app!

I notice that all the comments (and their immediate children) are fetched at once when item is first loaded, which can take some time if there are a lot of comments. Can we improve it so it only fetches when those comments are scrolled to (i.e. visible to users)?

The same can be applied to fetch stories as they come into view I believe (as opposed to 15 at a time right now, while my screen usually only displays 5 at a time).

**Comment by epishie** [comment]
Thanks for the comments!
Yes, it can definitely be improved.
I'll refactor the fetching logic to adapt into an infinite scroll setup instead of fetching by fixed chunks.

**Comment by epishie** [comment]
I've refactored the code to fetch items on demand.
2b5af79
3548850

I may have to fine tune some parameters though for smoother scrolling.

**Comment by hidroh** [comment]
Cool! Adding a dummy load more item that triggers auto load makes it load faster now, though hardcoding number of items to fetch to 5 makes the solution inflexible to different devices: e.g. on a 10 inch tablet, where about 10 items can be shown at a time, it loads 5 first, then 5 more after a while.

It also seems to recursively load the whole hierarchy now (instead of only immediate children), which is super cool! But on the other hand this handicaps your solution. E.g. if 1st comment has 500 descendants, it will take very long time just to load the 1st comment.

**Comment by epishie** [comment]
Yeah, I intend to pull the fetch size into resource value based on config (size + orientation).
With the comments tree, my plan is to add a "View more replies" placeholder for comments exceeding a certain amount of child items but I still have to figure out how to correctly insert the new children into the tree.

BTW, this is a nice exercise. I might eventually release it in the play store and create an iOS version :)

**Comment by hidroh** [comment]
For me the trick is to do a 2-pass layout of items, just like you need to do 2-pass network requests (1st one for the list of items, and then individual items). The 1st pass of layout will just render empty shells by their positions in the list, then the view binding mechanism will kick in and automagically help you decide which items to fetch & render in the 2nd pass.

This way you don't need to calculate how many items to fetch at a time, that would be the number of visible shells. Worst case is you happen to fetch just a little bit more, when actual contents are much longer than the shells. Also this can work efficiently with the new `RecyclerView.Adapter` API that allows you to notify individual item changes, as opposed to the old `ListView.Adapter` notifying all.

Of course it will be more tricky when you have hierarchy, because now you have to find out where to insert children (recursively) as well.
